File: services/game-api/internal/core/quest_side_effects.go
package core

// QuestSideEffects describes deterministic durable side effects derived only from
// a canonical MQ001..MQ100 quest identifier. It carries no player-supplied data.
type QuestSideEffects struct {
	InventoryItemID      string
	InventoryQuantity    int64
	BlueprintID          string
	MarkStarterRoadworthy bool
}

// QuestSideEffectsFor derives the authoritative quest side-effect schedule.
// PostgreSQL remains the durable authority that applies these effects transactionally.
func QuestSideEffectsFor(questID string) (QuestSideEffects, error) {
	if _, err := ParseQuestID(questID); err != nil {
		return QuestSideEffects{}, err
	}

	switch questID {
	case "MQ004":
		return QuestSideEffects{InventoryItemID: "part_brakes_track_i", InventoryQuantity: 1}, nil
	case "MQ005":
		return QuestSideEffects{BlueprintID: StarterRebuildBlueprint}, nil
	case "MQ009":
		return QuestSideEffects{InventoryItemID: "part_tires_street_i", InventoryQuantity: 1}, nil
	case "MQ012":
		return QuestSideEffects{MarkStarterRoadworthy: true}, nil
	default:
		return QuestSideEffects{}, nil
	}
}

File: src/zneondrive/quest_side_effects.py
"""Reference-only quest side-effect schedule for Python/Go parity tests."""

from __future__ import annotations

from dataclasses import dataclass

from .domain import STARTER_REBUILD_BLUEPRINT, parse_quest_id


@dataclass(frozen=True, slots=True)
class QuestSideEffects:
    inventory_item_id: str = ""
    inventory_quantity: int = 0
    blueprint_id: str = ""
    mark_starter_roadworthy: bool = False


def quest_side_effects(quest_id: str) -> QuestSideEffects:
    """Derive canonical quest side effects without granting production authority."""
    parse_quest_id(quest_id)
    if quest_id == "MQ004":
        return QuestSideEffects("part_brakes_track_i", 1)
    if quest_id == "MQ005":
        return QuestSideEffects(blueprint_id=STARTER_REBUILD_BLUEPRINT)
    if quest_id == "MQ009":
        return QuestSideEffects("part_tires_street_i", 1)
    if quest_id == "MQ012":
        return QuestSideEffects(mark_starter_roadworthy=True)
    return QuestSideEffects()
